cpp/src/arrow/array/util.cc

```diff
@@ -442,9 +442,13 @@ class NullArrayFactory {
     // First buffer is always null
     out_->buffers[0] = nullptr;
 
-    // Type codes are all zero, so we can use buffer_ which has had it's memory
-    // zeroed
     out_->buffers[1] = buffer_;
+    // buffer_ is zeroed, but 0 may not be a valid type code
+    if (type.type_codes()[0] != 0) {
+      ARROW_ASSIGN_OR_RAISE(out_->buffers[1], AllocateBuffer(buffer_->size(), pool_));
```

Comment:
`length_` rather than `buffer_->size()`? (can `buffer_` be larger because of the children or offsets?)